pkg/api/injectsnippetsfs.go

```diff
@@ -0,0 +1,103 @@
+package api
+
+import (
+	"bytes"
+	"fmt"
+	"html"
+	"io"
+	"io/fs"
+	"time"
+)
+
+type InjectSnippetsFS struct {
+	fs.FS
+	name    string
+	content string
+}
+
+func NewInjectIndexFS(fsys fs.FS, name string, marker string, snippets map[string]string) (fs.FS, error) {
+	if len(snippets) == 0 {
+		// no snippets, return the original marker
+		return fsys, nil
+	}
+
+	// read the content and inject snippets
+	f, err := fsys.Open(name)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = f.Close() }()
+	all, err := io.ReadAll(f)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", name, err)
+	}
+
+	// inject snippets code into content
+	codeSnippets := renderCodeSnippets(snippets)
+	contentBytes := bytes.Replace(all, []byte(marker), codeSnippets, 1)
+	return &InjectSnippetsFS{
+		FS:      fsys,
+		name:    name,
+		content: string(contentBytes),
+	}, nil
+}
+
+func renderCodeSnippets(snippets map[string]string) []byte {
+	var b bytes.Buffer
+	for k, v := range snippets {
+		_, _ = b.WriteString("<!-- snippet begin: " + html.EscapeString(k) + " -->\n")
```

Comment:
This works, but only because of a technicality.  HTML comments end with an explicit "`-->`"; the [exact details of what is allowed inside are complex](https://html.spec.whatwg.org/#comments) but quoting every "`<`" and "`>`" appears -- as a side effect -- to resolve them.  Maybe just strip all non-ASCII or punctuation?  Or maybe package html/template can do it for us?

Reply:
I prefer to encode the text as is without inject and get the user description here :)